Add a new system power off API

gmarull opened the pull request:
Add a new API to perform an immediate system power off: `sys_poweroff()`. 

Until now, this functionality has been implemented via the system power
management module, but in a clunky fashion. The way system PM works is
by defining some idle states in devicetree, that, given some properties
(e.g. minimal residency, exit latency, etc.) are automatically selected
when system goes to idle based on the expected next wake-up. However,
system off or shutdown is a power state that one typically wants to
control manually from the application because it implies state loss,
and in most cases, configuring some sort of wake-up source. So in
general, it is not desired to let the system enter this state
automatically. This led to the following stuff in-tree:

from `boards/arm/mimxrt595_evk/mimxrt595_evk_cm33.dts`:

```c
/*
 * Deep power-down mode is supported in this SoC through
 *  'PM_STATE_SOFT_OFF' state. There is no entry for this in device tree,
 *  user can call pm_state_force to enter this state.
 */
```

That is, state not being defined in devicetree so that PM subsystem
doesn't pick it automatically, but still implemented in in the PM hooks:

from `soc/arm/nxp_imx/rt5xx/power.c`, `pm_state_set()`:
```c
case PM_STATE_SOFT_OFF:
	set_deepsleep_pin_config();
	POWER_EnterDeepPowerDown(EXCLUDE_FROM_DEEP_POWERDOWN);
	break;
```

And to actually make use of this state, users had to do this kind of
abominations:

```c
pm_state_force(0u, &(struct pm_state_info){ PM_STATE_SOFT_OFF, 0, 0 });

/* Now we need to go sleep. This will let the idle thread runs and
 * the pm subsystem will use the forced state. To confirm that the
 * forced state is used, lets set the same timeout used previously.
 */
k_sleep(K_SECONDS(SLEEP_S));

printk("ERROR: System off failed\n");
while (true) {
	/* spin to avoid fall-off behavior */
}
```

In the case of Nordic, it also allows us to get rid of all PM hooks, resulting in ~1K of ROM savings on all builds.

gmarull committed:
lib: os: add support for system power off

Add a new API to perform an immediate system power off:
`sys_poweroff()`.

Until now, this functionality has been implemented via the system power
management module, but in a clunky fashion. The way system PM works is
by defining some idle states in devicetree, that, given some properties
(e.g. minimal residency, exit latency, etc.) are automatically selected
when system goes to idle based on the expected next wake-up. However,
system off is a power state that one typically wants to control manually
from the application because it implies state loss, and in most cases,
configuring some sort of wake-up source. So in general, it is not
desired to let the system enter this state automatically. This led to
the following stuff in-tree:

from `boards/arm/mimxrt595_evk/mimxrt595_evk_cm33.dts`:

```c
/*
 * Deep power-down mode is supported in this SoC through
 *  'PM_STATE_SOFT_OFF' state. There is no entry for this in device tree,
 *  user can call pm_state_force to enter this state.
 */
```

That is, state not being defined in devicetree so that PM subsystem
doesn't pick it automatically, but still implemented in in the PM hooks:

from `soc/arm/nxp_imx/rt5xx/power.c`, `pm_state_set()`:
```c
case PM_STATE_SOFT_OFF:
	set_deepsleep_pin_config();
	POWER_EnterDeepPowerDown(EXCLUDE_FROM_DEEP_POWERDOWN);
	break;
```

And to actually make use of this state, users had to do this kind of
abominations:

```c
pm_state_force(0u, &(struct pm_state_info){ PM_STATE_SOFT_OFF, 0, 0 });

/* Now we need to go sleep. This will let the idle thread runs and
 * the pm subsystem will use the forced state. To confirm that the
 * forced state is used, lets set the same timeout used previously.
 */
k_sleep(K_SECONDS(SLEEP_S));

printk("ERROR: System off failed\n");
while (true) {
	/* spin to avoid fall-off behavior */
}
```

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
soc: arm: nordic_nrf: add support for poweroff

Implement the z_sys_poweroff() hook and select HAS_POWEROFF.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
samples: boards: nrf: system_off: use sys_poweroff

Make use of the new sys_poweroff API to actually power off the system.
This lets us remove many hacks present in the sample.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
soc: arm: nordic_nrf: drop PM hooks

Nordic SoCs do all power management automagically when going to idle (ie
k_cpu_idle()). The only extra state, system off, is now handled via
sys_shutdown(), so there's no need to support the PM subsystem.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
soc: arm: nxp_imx: rt5xx: add support for power off

Implement support for sys_poweroff(). Code re-used from the PM SOFT_OFF
implementation.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
samples: boards: mimxrt595_evk_cm33: system_off: use sys_poweroff

Use the new sys_poweroff() API to power off the system.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
soc: arm: nxp_imx: rt5xx: drop SOFT_OFF

SOFT_OFF is now handled via sys_shutdown() API.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

carlocaione commented:
on `ARM64` we introduced the `pm_cpu_ops` to interface with PSCI, see: https://github.com/zephyrproject-rtos/zephyr/blob/ddd2bc94e2f4b51a3ac6f1f0e63e5665266c0f3f/include/zephyr/drivers/pm_cpu_ops.h#L55-L66

Do you think we can make the two co-exist or at least having the `pm_cpu_ops` as stardard fallback for `ARM64`?

mmahadevan108 commented:
Please delete `PM_STATE_SOFT_OFF` from `include/zephyr/pm/state.h`.
Please document this change in the release notes.

gmarull commented:
> Please delete `PM_STATE_SOFT_OFF` from `include/zephyr/pm/state.h`.

This can only be done after all current implementations are migrated.

> Please document this change in the release notes.

Will do, bit this is typically done before the release.

gmarull committed:
drivers: pm_cpu_ops: psci: provide sys_poweroff hook

Instead of implementing a custom power off API (pm_system_off),
implement the sys_poweroff hook, and indicate power off is supported by
selecting HAS_POWEROFF. Note that according to the PSCI specification
(DEN0022E), the SYSTEM_OFF operation does not return, however, an error
is printed and system is halted in case this occurs.

Note that the pm_system_off has also been deleted, from now on, systems
supporting PSCI should enable CONFIG_POWEROFF and call the standard
sys_poweroff() API.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

carlocaione commented on include/zephyr/drivers/pm_cpu_ops.h:
```diff
@@ -52,19 +52,6 @@ int pm_cpu_off(void);
- * @retval does not return on success, a negative errno otherwise
- * @retval -ENOTSUP If the operation is not supported
- */
-int pm_system_off(void);
```
I think `arch/arm64/core/fatal.c` is still using `pm_system_off`.

gmarull replied:
fixed

ceolin replied:
what is gonna happen with the rest of this API ? What is the plan ? Will we have `sys_shutdown()` and `pm_system_reset()` ? And the others API here ?

gmarull commented on arch/arm64/core/fatal.c:
```diff
@@ -350,7 +351,10 @@ FUNC_NORETURN void arch_system_halt(unsigned int reason)
 	ARG_UNUSED(reason);
 
 	(void)arch_irq_lock();
-	(void)pm_system_off();
+
+#ifdef CONFIG_SHUTDOWN
+	sys_shutdown();
+#endif /* CONFIG_SHUTDOWN */
```
btw, should arch_system_halt be calling shutdown? IIRC halt means to stop the system, but not cut the power.

carlocaione commented on drivers/pm_cpu_ops/pm_cpu_ops_psci.c:
```diff
@@ -66,19 +70,12 @@ int pm_cpu_on(unsigned long cpuid,
-	ret = psci_data.invoke_psci_fn(PSCI_0_2_FN_SYSTEM_OFF, 0, 0, 0);
-
-	return psci_to_dev_err(ret);
+	(void)psci_data.invoke_psci_fn(PSCI_0_2_FN_SYSTEM_OFF, 0, 0, 0);
```
Here, we should check anyway for the conduit to be set.

carlocaione replied:
yes, that's fine as well.

gmarull replied:
done

carlocaione reviewed (dismissed):
LGTM for the general idea and the `pm_cpu_ops` part.

fabiobaltieri reviewed (dismissed):
Nice! Always found the current power off API a bit messy, this makes a lot of sense.

fabiobaltieri commented on kernel/idle.c:
```diff
@@ -14,9 +14,21 @@
 LOG_MODULE_DECLARE(os, CONFIG_KERNEL_LOG_LEVEL);
 
+#ifdef CONFIG_SHUTDOWN
+static bool shutdown_sheduled;
```
Should this be an atomic? Maybe some weird cases on SMP or something, `z_sys_shutdown_schedule` could be called from wherever.

gmarull replied:
Hmm... actually on SMP you'll have N idle threads, so I guess we should wait until all CPUs go to idle before shutting down. Please check latest revision.

nashif reviewed (changes requested):
Let's pause for a bit, does this all make sense that we are adding a shutdown.h API to join reboot.h API, what is going to be next? reset.h API? It is worth looking into consolidating all of this.

carlocaione commented:
> Let's pause for a bit, does this all make sense that we are adding a shutdown.h API to join reboot.h API, what is going to be next? reset.h API? It is worth looking into consolidating all of this.

At this point if we are really going to rework this maybe we can merge and extend `pm_cpu_ops` API ?

gmarull commented:
> Let's pause for a bit, does this all make sense that we are adding a shutdown.h API to join reboot.h API, what is going to be next? reset.h API? It is worth looking into consolidating all of this.

This is just improving something that is currently unusable, ie

```c
pm_state_force(0u, &(struct pm_state_info){ PM_STATE_SOFT_OFF, 0, 0 });

/* Now we need to go sleep. This will let the idle thread runs and
 * the pm subsystem will use the forced state. To confirm that the
 * forced state is used, lets set the same timeout used previously.
 */
k_sleep(K_SECONDS(SLEEP_S));

printk("ERROR: System off failed\n");
while (true) {
	/* spin to avoid fall-off behavior */
}
```

Shutdown hooks are also present on other systems, e.g. on Linux, so please, let's not block without a solid reason behind. I'm fine discussing the implementation details, but I'd not want to enter into endless discussions to find the final/eternal solution, as this will just lead to frustration. This has nothing to do with reboot, it's about shutting the system down.

gmarull commented:
> > Let's pause for a bit, does this all make sense that we are adding a shutdown.h API to join reboot.h API, what is going to be next? reset.h API? It is worth looking into consolidating all of this.
> 
> At this point if we are really going to rework this maybe we can merge and extend `pm_cpu_ops` API ?

I'm not sure pm_cpu_ops API is the right place for shutdown, I see it more like a system-wide action (same as reset/reboot, for instance). I actually used `sys_shutdown` to keep a minimal consistency with `sys_reboot`. Both could probably be merged into a single header, but I didn't want to cause more trauma to people with another breaking change (header rename in this case).

mmahadevan108 commented:
> > Please document this change in the release notes.
> 
> Will do, bit this is typically done before the release.

Whats a good way to ensure we do not miss during release time. Add a `RELEASE_NOTES` label?

ceolin commented on subsys/pm/pm.c:
```diff
@@ -210,12 +210,11 @@ bool pm_system_suspend(int32_t ticks)
 
 #if defined(CONFIG_PM_DEVICE) && !defined(CONFIG_PM_DEVICE_RUNTIME_EXCLUSIVE)
-	if (atomic_sub(&_cpus_active, 1) == 1) {
+	if (atomic_get(&_cpus_active) == 0) {
```
TOCTOU ?  N idle threads can get here and have `_cpus_active == 0` and end up all  of them suspending devices. The decrement and check has to be atomic.

ceolin commented on subsys/pm/pm.c:
```diff
@@ -242,7 +241,7 @@ bool pm_system_suspend(int32_t ticks)
 	/* Wake up sequence starts here */
 #if defined(CONFIG_PM_DEVICE) && !defined(CONFIG_PM_DEVICE_RUNTIME_EXCLUSIVE)
-	if (atomic_add(&_cpus_active, 1) == 0) {
+	if (atomic_get(&_cpus_active) == 0) {
```
ditto

ceolin commented on kernel/idle.c:
```diff
@@ -41,6 +55,14 @@ void idle(void *unused1, void *unused2, void *unused3)
 	while (true) {
+#ifdef CONFIG_SHUTDOWN
+		if ((atomic_get(&shutdown_sheduled) == 1) &&
+		    (atomic_get(&_cpus_active) == 1)) {
```
you are not preventing other cores to wake up, so you can definitely have a race condition where another core can wake up and start things while you are shutting down or you never  get this condition true.

ceolin replied:
or I missed something here :)

ceolin commented:
> > Let's pause for a bit, does this all make sense that we are adding a shutdown.h API to join reboot.h API, what is going to be next? reset.h API? It is worth looking into consolidating all of this.
> 
> This is just improving something that is currently unusable, ie

That is really not a clean way, but the snippet is not fair, the intended behavior result is obtained with two calls:

```
 pm_state_force(0u, &(struct pm_state_info){ PM_STATE_SOFT_OFF, 0, 0 });
 k_sleep(K_SECONDS(SLEEP_S));
```

> 
> ```c
> pm_state_force(0u, &(struct pm_state_info){ PM_STATE_SOFT_OFF, 0, 0 });
> 
> /* Now we need to go sleep. This will let the idle thread runs and
>  * the pm subsystem will use the forced state. To confirm that the
>  * forced state is used, lets set the same timeout used previously.
>  */
> k_sleep(K_SECONDS(SLEEP_S));
> 
> printk("ERROR: System off failed\n");
> while (true) {
> 	/* spin to avoid fall-off behavior */
> }
> ```
> 
> Shutdown hooks are also present on other systems, e.g. on Linux, so please, let's not block without a solid reason behind. I'm fine discussing the implementation details, but I'd not want to enter into endless discussions to find the final/eternal solution, as this will just lead to frustration. This has nothing to do with reboot, it's about shutting the system down.

ceolin commented:
> > Let's pause for a bit, does this all make sense that we are adding a shutdown.h API to join reboot.h API, what is going to be next? reset.h API? It is worth looking into consolidating all of this.
> 
> At this point if we are really going to rework this maybe we can merge and extend `pm_cpu_ops` API ?



> > > Let's pause for a bit, does this all make sense that we are adding a shutdown.h API to join reboot.h API, what is going to be next? reset.h API? It is worth looking into consolidating all of this.
> > 
> > 
> > At this point if we are really going to rework this maybe we can merge and extend `pm_cpu_ops` API ?
> 
> I'm not sure pm_cpu_ops API is the right place for shutdown, I see it more like a system-wide action (same as reset/reboot, for instance). I actually used `sys_shutdown` to keep a minimal consistency with `sys_reboot`. Both could probably be merged into a single header, but I didn't want to cause more trauma to people with another breaking change (header rename in this case).

I agree, this looks more a system wide api. @carlocaione `pm_cpu_ops` really make sense in drivers ? we have similar API for intel adsp in SOC level. My point is shouldn't make more sense to have something aligned with what @gmarull is proposing here, for example `sys_cpu_off` and `sys_cpu_on` ? Also the `pm_` namespace is confusing with the pm subsystem.

ceolin reviewed (changes requested):
AFAIU there are several race conditions on SMP and we need more discussion API wise (related with pm_cpu_ops).

carlocaione commented:
> I agree, this looks more a system wide api. @carlocaione `pm_cpu_ops` really make sense in drivers ? we have similar API for intel adsp in SOC level. My point is shouldn't make more sense to have something aligned with what @gmarull is proposing here, for example `sys_cpu_off` and `sys_cpu_on` ? Also the `pm_` namespace is confusing with the pm subsystem.

No, I strongly disagree here. On ARM64 (and this is definitely not a SOC-specific thing) you need to have some kind of firmware that is usually running in EL3 (monitor) to manage the PM part of the CPU (usually `on`, `off`, `suspend`, `shutdown`, etc...). This firmware can be `PSCI` compliant and going through `SMCCC` or doing its own thing. This is not some kind of platform / SoC custom magic.

We do not want to have some kind of SoC specific stuff for this to introduce some copy-and-paste code all over the ARM64 boards when the boards are compliant to `PSCI`. Even Linux has an API for this called `cpu_operations`.

If you want to have the `sys_cpu_*` ops falling back onto the `pm_cpu_ops` be my guest, but this is as much as I am available to concede on this topic.

About the name I'm not particularly invested in that.

fabiobaltieri commented:
I like this being a simple core API being such a fundamental concept of any battery powered design. Maybe the header could live in `pm/`, but feels like very much like an `exit()` to me so a level above seems to make sense.

gmarull commented:
Last push: I've gone back to the original shape. For SMP, shutdown will be immediate with the latest patch set. After all, a proper shutdown mechanism didn't exist for SMP before this proposal (either the first idle thread would run the PM SOFT_OFF routine, ignoring other cores), or PM CPU ops API could be used, meaning immediate shutdown. I think it's better to wait for a proper SMP implementation rather than quickly coding a naive implementation just to have the false sense of supporting SMP properly. So to summarize, the current proposal is about a new API (this should not change) that solves a real problem for quite a few platforms. Shutdown sequence can be refined later.

carlocaione commented:
> Last push: I've gone back to the original shape. For SMP, shutdown will be immediate with the latest patch set. After all, a proper shutdown mechanism didn't exist for SMP before this proposal (either the first idle thread would run the PM SOFT_OFF routine, ignoring other cores), or PM CPU ops API could be used, meaning immediate shutdown. I think it's better to wait for a proper SMP implementation rather than quickly coding a naive implementation just to have the false sense of supporting SMP properly. So to summarize, the current proposal is about a new API (this should not change) that solves a real problem for quite a few platforms. Shutdown sequence can be refined later.

Just to give some context, `PSCI` states that `The caller must place all cores in a known state prior to the call` so I'm fine with this change.

ceolin commented:
> > I agree, this looks more a system wide api. @carlocaione `pm_cpu_ops` really make sense in drivers ? we have similar API for intel adsp in SOC level. My point is shouldn't make more sense to have something aligned with what @gmarull is proposing here, for example `sys_cpu_off` and `sys_cpu_on` ? Also the `pm_` namespace is confusing with the pm subsystem.
> 
> No, I strongly disagree here. On ARM64 (and this is definitely not a SOC-specific thing) you need to have some kind of firmware that is usually running in EL3 (monitor) to manage the PM part of the CPU (usually `on`, `off`, `suspend`, `shutdown`, etc...). This firmware can be `PSCI` compliant and going through `SMCCC` or doing its own thing. This is not some kind of platform / SoC custom magic.

It was more a question than a proposal, I don't how it works on arm64, I was checking if it was the same as on intel adsp.

> 
> We do not want to have some kind of SoC specific stuff for this to introduce some copy-and-paste code all over the ARM64 boards when the boards are compliant to `PSCI`. Even Linux has an API for this called `cpu_operations` (and `psci_operations`).
> 
> If you want to have the `sys_cpu_*` ops falling back onto the `pm_cpu_ops` be my guest, but this is as much as I am available to concede on this topic.
> 
> About the name I'm not particularly invested in that.

ok

ceolin commented:
> Last push: I've gone back to the original shape. For SMP, shutdown will be immediate with the latest patch set. After all, a proper shutdown mechanism didn't exist for SMP before this proposal (either the first idle thread would run the PM SOFT_OFF routine, ignoring other cores), or PM CPU ops API could be used, meaning immediate shutdown. I think it's better to wait for a proper SMP implementation rather than quickly coding a naive implementation just to have the false sense of supporting SMP properly. So to summarize, the current proposal is about a new API (this should not change) that solves a real problem for quite a few platforms. Shutdown sequence can be refined later.

`PM_SOFT_OFF` was basically triggered by targets that manually forced this state, so the application responsible to sync, put other cores idle and the last one off. While this transfer a lot of responsibilities to the application it hasn't a race condition (at least not known).

gmarull commented:
> > Last push: I've gone back to the original shape. For SMP, shutdown will be immediate with the latest patch set. After all, a proper shutdown mechanism didn't exist for SMP before this proposal (either the first idle thread would run the PM SOFT_OFF routine, ignoring other cores), or PM CPU ops API could be used, meaning immediate shutdown. I think it's better to wait for a proper SMP implementation rather than quickly coding a naive implementation just to have the false sense of supporting SMP properly. So to summarize, the current proposal is about a new API (this should not change) that solves a real problem for quite a few platforms. Shutdown sequence can be refined later.
> 
> `PM_SOFT_OFF` was basically triggered by targets that manually forced this state, so the application responsible to sync, put other cores idle and the last one off. While this transfer a lot of responsibilities to the application it hasn't a race condition (at least not known).

This basically means we do not support shutdown on SMP based systems. The current implementation doesn't have any race conditions either, as it will perform immediate action (so still up to the application to do any necessary prep work). In any case, users of SMP systems should be aware that shutdown is not well-supported on their platforms.

gregshue commented:
Wow. I'm glad I looked at it, because this does not cover what I expected of "system shutdown". Nothing should prevent end users from tying long duration operations into SYS_INIT() (e.g., reinitialize a file system, read the entire SEEPROM, perform a mechanical initialization).  Each may have an analogous de-initialization that needs to be performed. 

Where is the de-init() of each of the SYS_INIT() calls?

gmarull commented:
> Wow. I'm glad I looked at it, because this does not cover what I expected of "system shutdown". Nothing should prevent end users from tying long duration operations into SYS_INIT() (e.g., reinitialize a file system, read the entire SEEPROM, perform a mechanical initialization). Each may have an analogous de-initialization that needs to be performed.
> 
> Where is the de-init() of each of the SYS_INIT() calls?

This PR has nothing to do with de-init and SYS_INIT, please read carefully the PR text. Feel free to open a feature request.

gregshue commented:
> This PR has nothing to do with de-init and SYS_INIT ...

The system integrators and downstream subsystem developers like me want to NOT have to look at PR requests to understand the scope of a function. 

What I noticed was the naming of the new file (sys/shutdown.h) and new function (sys_shutdown()), made me expect this functionality was a counterpart to SYS_INIT(). I am not confused by the name of the replaced function (`pm_system_off()`).  Do you have a more apparent name for the counterpart to SYS_INIT()?  If not, then consider renaming sys_shutdown to something that more closely indicates the limited scope of functionality. ;-)

gmarull commented:
> > This PR has nothing to do with de-init and SYS_INIT ...
> 
> The system integrators and downstream subsystem developers like me want to NOT have to look at PR requests to understand the scope of a function.
> 
> What I noticed was the naming of the new file (sys/shutdown.h) and new function (sys_shutdown()), made me expect this functionality was a counterpart to SYS_INIT(). I am not confused by the name of the replaced function (`pm_system_off()`). Do you have a more apparent name for the counterpart to SYS_INIT()? If not, then consider renaming sys_shutdown to something that more closely indicates the limited scope of functionality. ;-)

I can only suggest to read the PR messages more carefully next time :-)

anangl commented on include/zephyr/sys/shutdown.h:
```diff
@@ -0,0 +1,38 @@
+/*
+ * Copyright (c) 2015 Wind River Systems, Inc.
```
Incorrect copyright notice, here and in `shutdown_internal.h`.

gmarull replied:
fixed

carlescufi commented:
Architecture WG:

- @gmarull introduces the subystem: This is a clean shutdown API that does not require the power management subsystem. The power management subsystem tends to be opinionated, whereas this subsystem is fully controlled by the application. Similar to `PM_STATE_FORCE` but also blocking.
- @bjarki-trackunit likes the solution, also has the same usecase
- @nashif asks what "shutdown" means in this context. @gmarull replies that this is "shutting down and never come back". Any actions required to wake up must be done by the application before calling shutdown
- @nashif asks whether we really need a new header file instead of trying to consolidate these APIs, just like `reboot` as well
- @carlescufi suggests that we could perhaps have an `include/sys/system.h` with APIs that are similar: reboot, shutdown.

gmarull commented:
last push: rebased to fix conflicts

fabiobaltieri commented:
Drop [RFC] from the title?

>     * @carlescufi suggests that we could perhaps have an `include/sys/system.h` with APIs that are similar: reboot, shutdown

I think this was the main actionable point from the discussion, are you planning on following up on that separately? Does it makes sense to have this as header just to include other `sys/` headers?

gmarull commented:
> Drop [RFC] from the title?
> 
> > ```
> > * @carlescufi suggests that we could perhaps have an `include/sys/system.h` with APIs that are similar: reboot, shutdown
> > ```
> 
> I think this was the main actionable point from the discussion, are you planning on following up on that separately? 

I don't really like `sys/system.h` (double `sys`), but I'm open to suggestions here. I don't see current header a problem either, on Linux you have `sys/reboot.h` with a single function as well (main difference is that `reboot` takes an argument that can be used to perform shutdown: `RB_POWER_OFF`). 

> Does it makes sense to have this as header just to include other `sys/` headers?

`sys/` is really heterogeneous to make a catch-all header, and I've personally never been a fan of such headers.

gmarull commented:
After some thinking, keeping shutdown API in its own header makes sense because it is an independent feature (CONFIG_SHUTDOWN). It also has its own internal header (shutdown_internal.h). For reference, on Linux there's sys/reboot.h with a single function as well (note that on Linux `reboot` can be used for multiple things like reboot, shutdown, etc.)

Thoughts? Other naming suggestions?

carlocaione reviewed (dismissed):
I won't fight over naming, LGTM.

fabiobaltieri commented:
> Thoughts? Other naming suggestions?

I don't mind separate header, in how many files of the application do you expect that to end up? Probably somewhere between 0 and 1.

ceolin commented on lib/os/shutdown.c:
```diff
@@ -0,0 +1,22 @@
+{
+#if defined(CONFIG_MULTITHREADING) && !defined(CONFIG_SMP)
+	z_sys_shutdown_schedule();
+	k_sleep(K_FOREVER);
```
```
pm_state_force(0u, &(struct pm_state_info){ PM_STATE_SOFT_OFF, 0, 0 });
k_sleep(K_FOREVER);
```

Basically ...

gmarull replied:
yes, but just with `sys_shutdown()` call, so:

- No clunky inline structs (what are those 0?)
- No need to sleep outside (K_FOREVER? Some value and later a fallback loop "just in case it fails"...?)
- API gives you guarantees of the behavior, so you can write portable code
- No need to do hacks like avoiding soft-off in devicetree (just see comments in NXP files, or the recently removed em4 in silabs)
- No need to do hacks like `disable_ds_1` in Nordic system off sample

To me, this really makes a difference. If for some it doesn't, I guess I'm done with the PR.

ceolin replied:
that is not what I meant, from user the API simplify things no doubt. Talking about the internal implementation that could be achieved doing that.

> K_FOREVER? Some value and later a fallback loop "just in case it fails"...?)
```
printk("ERROR: System off failed\n");
while (true) {
	/* spin to avoid fall-off behavior */
}
```
Could easily be replaced with `CODE_UNREACHABLE;` as it is done in ` sys_shutdown`. Just pointing that this pattern is similar.

> No need to do hacks like avoiding soft-off in devicetree (just see comments in NXP files, or the recently removed em4 in silabs)

That is a good point indeed.

ceolin commented:
I keep the -1 not because the API itself, but because this needs more discussion around the whole picture.

- `pm_system_off` has gone and the new `sys_shutdown` was included, though, we still have `pm_system_reset` ... What is the plan ? It is confuse, these functionalities should follow some similar namespace convention.
- This is a feature that is related with power management (even if it is available for targets that don't have pm). I think it should have some integration with the power management subsystem when it is enabled. I think we will end up replicating logic when dealing with SMP. Also for targets that are using system power management, it makes sense the shutdown following the system power management flow and suspending devices.
-   For targets with SOFT_OFF and shutdown enabled, do they need to duplicate the implementation ?

gmarull commented:
> I keep the -1 not because the API itself, but because this needs more discussion around the whole picture.
> 
> * `pm_system_off` has gone and the new `sys_shutdown` was included, though, we still have `pm_system_reset` ... What is the plan ? It is confuse, these functionalities should follow some similar namespace convention.

I guess the question is why `pm_system_reset` was introduced out of `sys_reboot` (https://github.com/zephyrproject-rtos/zephyr/commit/62dbe72cb7ac55d1ae3ce62e5fb1fae47b9b293a). It's basically a clone of an existing API. Anything to say here?
 
> * This is a feature that is related with power management (even if it is available for targets that don't have pm). I think it should have some integration with the power management subsystem when it is enabled. I think we will end up replicating logic when dealing with SMP.

First, SMP + shutdown is not supported properly today. Neither with PM/pm_cpu_ops API. There's no need to duplicate code if a future implementation finds commonalities between shutdown/PM, just use some shared code.

> Also for targets that are using system power management, it makes sense the shutdown following the system power management flow and suspending devices.

It doesn't on systems using device PM runtime. I'd like to emphasize again that PM_DEVICE has multiple issues, to the point it's broken for many use-cases, so it should be avoided to build something solid and stable (or re-implemented properly). I have yet to hear real use cases that can't be solved in a better way.

> * For targets with SOFT_OFF and shutdown enabled, do they need to duplicate the implementation ?

sys_shutdown() is a replacement for SOFT_OFF. This PR removes SOFT_OFF from targets implementing sys_shutdown().

carlocaione commented:
> * `pm_system_off` has gone and the new `sys_shutdown` was included, though, we still have `pm_system_reset` ... What is the plan ? It is confuse, these functionalities should follow some similar namespace convention.

Again, why this keeps being put in the mix? The whole `include/zephyr/drivers/pm_cpu_ops.h` was introduced as a formalized API / interface for `PSCI` (well defined standard) and in general for firmware running outsize Zephyr for PM on ARM64 platforms, it was never meant to be used as a **general** interface for all the Zephyr platforms. We can bridge the two, but for me those are two different separated things.

gmarull commented:
> > * `pm_system_off` has gone and the new `sys_shutdown` was included, though, we still have `pm_system_reset` ... What is the plan ? It is confuse, these functionalities should follow some similar namespace convention.
> 
> Again, why this keeps being put in the mix? The whole `include/zephyr/drivers/pm_cpu_ops.h` was introduced as a formalized API / interface for `PSCI` (well defined standard) and in general for firmware running outsize Zephyr for PM on ARM64 platforms, it was never meant to be used as a **general** interface for all the Zephyr platforms. We can bridge the two, but for me those are two different separated things.

So apparently it's now mixed with `sys_reboot`... `pm_system_reset` takes arguments mirrored from `sys/reboot.h`. And interestingly enough, it accepts `SYS_COLD_RESET` even though not part of the spec:

![image](https://github.com/zephyrproject-rtos/zephyr/assets/25011557/5b222cfe-8a87-4d0a-9444-a632f0da29e6)

This PR removes `pm_system_off`, but I guess it could be kept and called in the PSCI impl of `sys_shutdown`.

carlescufi commented:
> * `pm_system_off` has gone and the new `sys_shutdown` was included, though, we still have `pm_system_reset` ... What is the plan ? It is confuse, these functionalities should follow some similar namespace convention.

The problem, as stated already by @gmarull, is that `pm_system_reset` should not exist at all. In fact, I've created an issue to remove it:
https://github.com/zephyrproject-rtos/zephyr/issues/60915

This should hopefully clear out the confusion you mention.

> * This is a feature that is related with power management (even if it is available for targets that don't have pm). I think it should have some integration with the power management subsystem when it is enabled. I think we will end up replicating logic when dealing with SMP. Also for targets that are using system power management, it makes sense the shutdown following the system power management flow and suspending devices.

I see no issue in handling shutdowns via a separate interface (and in fact no one else did during the discussion in the architecture WG meeting) and still share any common code with PM when applicable. This can be achieved by refactoring as necessary. Linux also follows this model where reboot and shutdown are not directly tied to the PM power states, both for the [reboot functionality](https://elixir.bootlin.com/linux/v6.4.4/source/include/linux/reboot.h) and the [shutdown one](https://elixir.bootlin.com/linux/v6.4.4/source/include/linux/pm.h#L23).

> * For targets with SOFT_OFF and shutdown enabled, do they need to duplicate the implementation ?

No, because the idea is to migrate all platforms to `sys_shutdown` and get rid of `SOFT_OFF`.

ceolin commented:
> > * `pm_system_off` has gone and the new `sys_shutdown` was included, though, we still have `pm_system_reset` ... What is the plan ? It is confuse, these functionalities should follow some similar namespace convention.
> 
> Again, why this keeps being put in the mix? The whole `include/zephyr/drivers/pm_cpu_ops.h` was introduced as a formalized API / interface for `PSCI` (well defined standard) and in general for firmware running outsize Zephyr for PM on ARM64 platforms, it was never meant to be used as a **general** interface for all the Zephyr platforms. We can bridge the two, but for me those are two different separated things.

I never said they are the same my problem is with API consistency

ceolin commented:
> > I keep the -1 not because the API itself, but because this needs more discussion around the whole picture.
> > 
> > * `pm_system_off` has gone and the new `sys_shutdown` was included, though, we still have `pm_system_reset` ... What is the plan ? It is confuse, these functionalities should follow some similar namespace convention.
> 
> I guess the question is why `pm_system_reset` was introduced out of `sys_reboot` ([62dbe72](https://github.com/zephyrproject-rtos/zephyr/commit/62dbe72cb7ac55d1ae3ce62e5fb1fae47b9b293a)). It's basically a clone of an existing API. Anything to say here?
> 
> > * This is a feature that is related with power management (even if it is available for targets that don't have pm). I think it should have some integration with the power management subsystem when it is enabled. I think we will end up replicating logic when dealing with SMP.
> 
> First, SMP + shutdown is not supported properly today. Neither with PM/pm_cpu_ops API. There's no need to duplicate code if a future implementation finds commonalities between shutdown/PM, just use some shared code.
> 
> > Also for targets that are using system power management, it makes sense the shutdown following the system power management flow and suspending devices.
> 
> It doesn't on systems using device PM runtime. I'd like to emphasize again that PM_DEVICE has multiple issues, to the point it's broken for many use-cases, so it should be avoided to build something solid and stable (or re-implemented properly). I have yet to hear real use cases that can't be solved in a better way.

And I am gonna tell you again, that model may not be the correct for everyone and that is fine they can use `DEVICE_RUNTIME` but that is something that is working for some targets (Intel ADSP among others). They decided to take this route because drivers can't be suspended due latency and when the main CPU (that controls the DSP power management) asks the DSP to suspend they rely on system PM to suspend their devices. I tell you, that is a real use case, don't assume that you are aware of all use cases, that is a big mistake when starting removing/changing everything.

> 
> > * For targets with SOFT_OFF and shutdown enabled, do they need to duplicate the implementation ?
> 
> sys_shutdown() is a replacement for SOFT_OFF. This PR removes SOFT_OFF from targets implementing sys_shutdown().

nashif commented:
> After some thinking, keeping shutdown API in its own header makes sense because it is an independent feature (CONFIG_SHUTDOWN). It also has its own internal header (shutdown_internal.h). For reference, on Linux there's sys/reboot.h with a single function as well (note that on Linux `reboot` can be used for multiple things like reboot, shutdown, etc.)
> 
> Thoughts? Other naming suggestions?

folding shutdown in reboot and expanding reboot API similar to linux is a good way to move forward. Take this oppertunity to expand the API with reason for operation, If I remember correctly, this was a point of discussion in the past and useful when a system comes back (i.e. by using retention).

APIs for shutdown, reboot available as general interfaces, independent of PM being enable or not is good to have, but looking at this disucssion it seems like we have some existing inconsistency and differnent ways of doing all of this, so it would be could to unify some of the code paths and clean up duplicate APIs along they way.

ceolin commented:
> > * `pm_system_off` has gone and the new `sys_shutdown` was included, though, we still have `pm_system_reset` ... What is the plan ? It is confuse, these functionalities should follow some similar namespace convention.
> 
> The problem, as stated already by @gmarull, is that `pm_system_reset` should not exist at all. In fact, I've created an issue to remove it: #60915
> 
> This should hopefully clear out the confusion you mention.
> 
> > * This is a feature that is related with power management (even if it is available for targets that don't have pm). I think it should have some integration with the power management subsystem when it is enabled. I think we will end up replicating logic when dealing with SMP. Also for targets that are using system power management, it makes sense the shutdown following the system power management flow and suspending devices.
> 
> I see no issue in handling shutdowns via a separate interface (and in fact no one else did during the discussion in the architecture WG meeting) and still share any common code with PM when applicable. This can be achieved by refactoring as necessary. Linux also follows this model where reboot and shutdown are not directly tied to the PM power states, both for the [reboot functionality](https://elixir.bootlin.com/linux/v6.4.4/source/include/linux/reboot.h) and the [shutdown one](https://elixir.bootlin.com/linux/v6.4.4/source/include/linux/pm.h#L23).
> 
> > * For targets with SOFT_OFF and shutdown enabled, do they need to duplicate the implementation ?
> 
> No, because the idea is to migrate all platforms to `sys_shutdown` and get rid of `SOFT_OFF`.

Why get rid of SOFT_OFF ? My understanding is that `sys_shutdown` would be something equivalent to Mechanical Off (ACPI G3) some devices may take different approaches for shutdown and SOFT_OFF. I pointed it out because in the replaced cases they were doing the same thing, but there is semantic difference.

ceolin commented:
> > > I keep the -1 not because the API itself, but because this needs more discussion around the whole picture.
> > > 
> > > * `pm_system_off` has gone and the new `sys_shutdown` was included, though, we still have `pm_system_reset` ... What is the plan ? It is confuse, these functionalities should follow some similar namespace convention.
> > 
> > 
> > I guess the question is why `pm_system_reset` was introduced out of `sys_reboot` ([62dbe72](https://github.com/zephyrproject-rtos/zephyr/commit/62dbe72cb7ac55d1ae3ce62e5fb1fae47b9b293a)). It's basically a clone of an existing API. Anything to say here?
> > > * This is a feature that is related with power management (even if it is available for targets that don't have pm). I think it should have some integration with the power management subsystem when it is enabled. I think we will end up replicating logic when dealing with SMP.
> > 
> > 
> > First, SMP + shutdown is not supported properly today. Neither with PM/pm_cpu_ops API. There's no need to duplicate code if a future implementation finds commonalities between shutdown/PM, just use some shared code.
> > > Also for targets that are using system power management, it makes sense the shutdown following the system power management flow and suspending devices.
> > 
> > 
> > It doesn't on systems using device PM runtime. I'd like to emphasize again that PM_DEVICE has multiple issues, to the point it's broken for many use-cases, so it should be avoided to build something solid and stable (or re-implemented properly). I have yet to hear real use cases that can't be solved in a better way.
> 
> And I am gonna tell you again, that model may not be the correct for everyone and that is fine they can use `DEVICE_RUNTIME` but that is something that is working for some targets (Intel ADSP among others). They decided to take this route because drivers can't be suspended due latency and when the main CPU (that controls the DSP power management) asks the DSP to suspend they rely on system PM to suspend their devices. I tell you, that is a real use case, don't assume that you are aware of all use cases, that is a big mistake when starting removing/changing everything.

Let me work through this, making it properly optional and letting the target choose which states should trigger it.

> 
> > > * For targets with SOFT_OFF and shutdown enabled, do they need to duplicate the implementation ?
> > 
> > 
> > sys_shutdown() is a replacement for SOFT_OFF. This PR removes SOFT_OFF from targets implementing sys_shutdown().

ceolin commented:
> > > No, because the idea is to migrate all platforms to `sys_shutdown` and get rid of `SOFT_OFF`.
> > 
> > 
> > Why get rid of SOFT_OFF ? My understanding is that `sys_shutdown` would be something equivalent to Mechanical Off (ACPI G3) some devices may take different approaches for shutdown and SOFT_OFF. I pointed it out because in the replaced cases they were doing the same thing, but there is semantic difference.
> 
> Well, `sys_shutdown` is a shutdown, not a power off. You can wake up from shutdown without user interaction on some platforms (eg. external GPIO stimulus). Regardless, we can keep `SOFT_OFF` independently of this new API.
> 
> What is exactly blocking this PR at this point @ceolin? It's not clear to me at this point. The API makes sense and there's wide agreement on it, we also agreed that the overlap you mentioned will be fixed, in part by @gmarull's #60920.

@carlescufi Nothing from my side now. Thanks @gmarull

andyross commented on lib/os/shutdown.c:
```diff
@@ -0,0 +1,22 @@
+	z_sys_shutdown_schedule();
+	k_sleep(K_FOREVER);
+#else
+	z_sys_shutdown();
```
This (in the platform implements I see here on NXP and Nordic) is synchronous, and will effect a shutdown immediately regardless of whatever the rest of the system is doing.

andyross reviewed (changes requested):
Sorry, late to the game, and apologies for not getting the -1 in earlier.

Maybe I'm having trouble understanding the goal here, but right now we seem to have two shutdown idioms in the same PR, with rather different semantics.  And nothing seems to be documenting the difference?

Why bother having a standard shutdown API if it isn't going to be rigorous about what it means to "shut down"?  All we're getting here is a standard name!

If you have an app or library on, I dunno, RISC-V that expects to use sys_shutdown() to do a lazy shutdown and wait for cleanup code to finish (which looks like it would work), and port to nRF, it's suddenly going to fail because the cleanup code doesn't run.

Can we pick just one?  Or am I misunderstanding the design?

andyross commented:
> Which two idioms? This introduces a single new API, aiming to replace pm_state_force + SOFT_OFF, and for now, it keeps the previous behavior.

Yeah... that's worse and not better, IMHO.  We used to have a simple API that did one thing on some platforms, and a complicated API that did a subtly different thing on different platforms.  Now we have the same call you can use on both platforms, but it still does different things?!

I just don't get it.  What's the advantage to merging this?  Won't it make things harder when we decide that sys_shutdown() needs properly-documented semantics because it will be a breaking change?

Frankly this seems cart-before-the-horse to me.  What we clearly want is a nicely specified shutdown/power-off API that tells the user exactly what they're getting and under what circumstances.  And once we have that we can make it pretty.

gmarull commented:
> > Which two idioms? This introduces a single new API, aiming to replace pm_state_force + SOFT_OFF, and for now, it keeps the previous behavior.
> 
> Yeah... that's worse and not better, IMHO. We used to have a simple API that did one thing on some platforms, and a complicated API that did a subtly different thing on different platforms. Now we have the same call you can use on both platforms, but it still does different things?!
> 

Well, the idea is to convert remaining platforms, not to keep multiple different ways. We actually have 2 today (PM path and pm_cpu_ops path).

> I just don't get it. What's the advantage to merging this? Won't it make things harder when we decide that sys_shutdown() needs properly-documented semantics because it will be a breaking change?

One of the advantages is to offer an API that offers some minimal guarantees (like for example that once called it will never return) and to be decoupled from PM, which in the case of Nordic it allows us to save a bit less than ~1K of ROM on all builds.

> 
> Frankly this seems cart-before-the-horse to me. What we clearly want is a nicely specified shutdown/power-off API that tells the user exactly what they're getting and under what circumstances. And once we have that we can make it pretty.

I think this PR addresses some existing problems, other features can be added incrementally. Also, this change is not breaking anything (as soon as we keep the definition of SOFT_OFF in PM). If in ~8y no one has shown interest in implementing a more complex shutdown like e.g. Linux, it's probably because it's not needed or because there's not enough knowledge/man power to do it.

andyross commented:
I think we're not really getting each others points?  Stated simply, my position is "No new APIs with inconsistent semantics, period.".  Maybe you could add a sys_shutdown_async() if you really feel we need a function here?  I wouldn't stand in the way there.

And the glitch here isn't remotely theoretical.  Consider a PM-enabled system handling a fatal error of some kind.   It does `LOG_ERR("The battery is on fire, state dump:"); ...; sys_shutdown();` and the recovery apparatus comes along later and uploads the telemetry.  Now you port the same system to Nordic and the **telemetry is lost**, because the shutdown is synchronous.  The previous system worked because the log backend got a chance to flush before shutdown.

Saying things like "well, the system is going to shut down at some point in the future anyway, so what's the difference" makes us look like we don't understand why we have the API in the first place.  I 100% agree that this is an area with complicated platform-dependencies and widely varying requirements.  That should make us **more** hesitant to wrap it in a pretty API, not less.

carlescufi commented:
> > Which two idioms? This introduces a single new API, aiming to replace pm_state_force + SOFT_OFF, and for now, it keeps the previous behavior.
> 
> Yeah... that's worse and not better, IMHO. We used to have a simple API that did one thing on some platforms, and a complicated API that did a subtly different thing on different platforms. Now we have the same call you can use on both platforms, but it still does different things?!

@andyross I am a bit lost myself here as well. Could you clarify to which two idioms you were referring as well as what the "simple API" and "complicated API" are? It'd help if you referred them by name since we seem to be missing info here in order to understand your points.

carlescufi commented on lib/os/shutdown.c:
```diff
@@ -0,0 +1,22 @@
+
+void sys_shutdown(void)
+{
+#if defined(CONFIG_MULTITHREADING) && !defined(CONFIG_SMP)
```
I think I finally understand what @andyross meant and I agree that the inconsistency between SMP and single-processor is not something we want to introduce. On top of that, `sys_reboot()` is fully synchronous and immediate, so I would vote for this API to become a synchronous, immediate one as well for consistency.

gmarull commented:
Ok guys, here's my new take on this PR:

- Action is now immediate, idle thread is no longer involved. The reason the previous solution involved idle thread was to preserve `pm_state_force` behavior, but eventually, it complicated things for no obvious reason.
- Current approach mimics what `sys_reboot()` does, i.e., an immediate action
- I renamed the function to sys_poweroff, I think it makes things less confusing as shutdown may be interpreted as a kind of _ordered_ shutdown.
- If in the future we support a proper shutdown sequence, the existing reboot/power off hooks will likely be useful as a last step in the chain. For now, it's application responsibility to make sure it is safe to call the API.

bjarki-andreasen reviewed (dismissed):
This fits my usecase well :)

nashif commented:
@gmarull  you often complain about missing documentation and this one is lacking documentation, so please add some docs.

Also, there is one thing I dislike here which is the use of z_ for the hook, this is reserved for internal APIs that should never be used outside of the Zephyr tree and zephyr code, so it would be beneficial to to come up with a naming convention for such hooks (also for the reboot PR) that makes it clear those are documented interfaces that can be used out of tree. For example `sys_poweroff_hook()` instead of `z_sys_poweroff()` or something along those lines. Such APIS will be part of another class of interfaces we need to document and maintain, but will have a different life cycle and change policy (unlike stable APIs).

gmarull commented:
Last push: added a page that imports `sys_poweroff` doxygen group.

I won't change the hook name, as the current convention is to use `z_` prefix for internal functions. I never liked this convention, but I prefer consistency to random rules. If this is the only blocker for this PR, I prefer to close it.

carlescufi reviewed (approved):
LGTM now, thanks @gmarull for the patience!

bjarki-andreasen reviewed (approved):
Looking forward to using the new API :)

carlocaione reviewed (approved):
LGTM. We can poweroff this PR.

andyross reviewed (approved):
Semantics are consistent and documented, no complaints here.  Thanks!

nashif commented:
> Last push: added a page that imports `sys_poweroff` doxygen group.

and that is what I was asking for, it was not there when I reviewed last. 

> I won't change the hook name, as the current convention is to use `z_` prefix for internal functions (hooks are just internal functions). I never liked this convention, but I prefer consistency to random rules. If this is the only blocker for this PR, I prefer to close it.

Ok, my bad that I mixed whatever is happening here with the reboot PR. This PR does not implement poweroff on different levels (arch -> soc -> ...) and does not have the notion of custom poweroff similar to what we have in the reboot PR where a custom implementation or hook can be implemented to override architecture code. So usage of internal API and naming is fine. 

The discussion on hooks vs internal APIs can move to the reboot PR. Sorry about the confusion.
